src/NuGet.Core/Microsoft.Build.NuGetSdkResolver/GlobalJsonReader.cs

```diff
@@ -20,60 +21,72 @@ internal static class GlobalJsonReader
                 return null;
             }
 
             DirectoryInfo projectDirectory = Directory.GetParent(context.ProjectFilePath);
 
-            if (!TryGetPathOfFileAbove(GlobalJsonFileName, projectDirectory, out FileInfo globalJsonPath))
+            if (projectDirectory == null || !TryGetPathOfFileAbove(fileName, projectDirectory, out FileInfo globalJsonPath))
             {
                 return null;
             }
 
-            string contents = File.ReadAllText(globalJsonPath.FullName);
-
-            // Look ahead in the contents to see if there is an msbuild-sdks section.  Deserializing the file requires us to load
-            // Newtonsoft.Json which is 500 KB while a global.json is usually ~100 bytes of text.
-            if (contents.IndexOf(MSBuildSdksPropertyName, StringComparison.Ordinal) == -1)
+            if (!globalJsonPath.Exists)
             {
-                return null;
-            }
+                FileCache.TryRemove(globalJsonPath, out var _);
 
-            try
-            {
-                return Deserialize(contents);
-            }
-            catch (Exception e)
-            {
-                // Failed to parse "{0}". {1}
-                var message = string.Format(CultureInfo.CurrentCulture, Strings.FailedToParseGlobalJson, globalJsonPath, e.Message);
-                context.Logger.LogMessage(message);
                 return null;
             }
-        }
 
-        /// <summary>
-        /// Deserializes a global.json and returns the MSBuild SDK versions
-        /// </summary>
-        [MethodImpl(MethodImplOptions.NoInlining)]
-        private static Dictionary<string, string> Deserialize(string value)
-        {
-            return JsonConvert.DeserializeObject<GlobalJsonFile>(value).MSBuildSdks;
-        }
+            // Add a new file to the cache if it doesn't exist.  If the file is already in the cache, read it again if the file has changed
+            (DateTime LastWriteTime, Lazy<Dictionary<string, string>> Lazy) result = FileCache.AddOrUpdate(
+                globalJsonPath,
+                key => (key.LastWriteTime, new Lazy<Dictionary<string, string>>(() =>
+                {
+                    wasRead = true;
 
-        private sealed class GlobalJsonFile
-        {
-            [JsonProperty(MSBuildSdksPropertyName)]
-            public Dictionary<string, string> MSBuildSdks { get; set; }
+                    return ParseMSBuildSdkVersions(key, context);
+                })),
+                (key, item) =>
+                {
+                    DateTime lastWriteTime = key.LastWriteTime;
+
+                    if (item.LastWriteTime < lastWriteTime)
+                    {
+                        return (lastWriteTime, new Lazy<Dictionary<string, string>>(() =>
+                        {
+                            wasRead = true;
```

Comment:
As far as I can tell, `wasRead` and `wasGlobalJsonRead` are only used for testing. Which is ok, except that MSBuild SDK resolving is on the hot path for project load (both in VS and command line).  The reason I bring up `wasRead` is because it's captured scope from the `GetMSBuildSdkVersions` method, and anonymous methods cause allocations. In this case, there's one anonymous method passed to `Lazy<T>`, and allocating the `Lazy<T>` itself is inside another anonymous method that is passed to `ConcurrentDictionary.AddOrUpdate`.

If `wasRead` is removed, then I believe that these methods could then be static and moved to class methods, allowing the delegates passed to `GetOrAdd` and `Lazy<T>`'s constrctor to be zero allocation.

I'll understand if some people consider the perf impact acceptable to validate correctness. I just wanted to make sure this was considered since `wasRead`/`wasGlobalJsonRead` is not used in production code.

Reply:
This is a good point.  Adding testability at the sacrifice of performance here probably isn't great.  What do you think about including this only for Debug bits?  For example:

```c#
#if DEBUG
   wasRead = true;
#endif
```

I'm not sure if our builds that run tests only build Debug.  I could `ifdef` the tests out on Release bits I suppose.  What does everybody think?

Reply:
Even if you don't capture scope, simply using an anonymous method will still cause allocations (unless the C# compiler has improvements I'm not aware of). So, using `#if` to stop using captured scope, but still using the anonymous method, then we won't get any perf benefit.

I believe the only way to remove allocations when passing delegates is to use a named class method (I hope I'm wrong)

Reply:
I've decided to use an interface to make the class more testable and used an event instead which fires if global.json is read.  Let me know what you think!